### PR TITLE
WIP: forward(), yet again.

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
@kaiidams, @lostmsu -- here's another version. This adds interfaces to Katsuya's version. The interface is useful for Sequential composition. It also helps make ScriptModule more general.

### CODE REVIEW

File: src/TorchSharp/JIT/ScriptModule.cs [hunk 1]
```diff
@@ -268,91 +269,102 @@ private Type GetType(Type type)
...
+
+
+                    switch (typeCode) {
+                    case 0:
+                        // Nothing.
+                        throw new NotImplementedException("ScriptModule.forward() returning something else than a tensor, a tuple of tensors, or list of tensors.");
+                    case 1:
+                        // Tensor
+                        return new Tensor(ptrArray[0]);
+                    case 2:
```
Comment (lostmsu) [inline src/TorchSharp/JIT/ScriptModule.cs:303]: Can C++ modules return a tuple of 0 elements?

File: src/TorchSharp/NN/Activation/ELU.cs [hunk 2]
```diff
@@ -12,12 +12,12 @@ namespace Modules
...
         /// This class is used to represent a ELU module.
         /// </summary>
-        public class ELU : torch.nn.Module
+        public class ELU : torch.nn.Module<Tensor, Tensor>
         {
             internal ELU(IntPtr handle, IntPtr boxedHandle) : base(handle, boxedHandle) { }
 
             [DllImport("LibTorchSharp")]
-            private static extern IntPtr THSNN_ELU_forward(torch.nn.Module.HType module, IntPtr tensor);
+            private static extern IntPtr THSNN_ELU_forward(torch.nn.Module<Tensor, Tensor>.HType module, IntPtr tensor);
```
Comment (lostmsu) [inline src/TorchSharp/NN/Activation/ELU.cs:20]: Does every generic module type need its own `HType`?

File: src/TorchSharp/NN/Losses.cs [hunk 3]
```diff
@@ -12,16 +12,56 @@ namespace TorchSharp
...
+    {
+        public WeightedLoss(Tensor? weight = null, torch.nn.Reduction reduction = nn.Reduction.Mean) : base(reduction)
+        {
+            this.weight = weight;
+        }
+
+        public Tensor? weight { get; }
+    }
+
+    public abstract class WeightedLoss<T1, T2, T3, TResult> : Loss<T1, T2, T3, TResult>
```
Comment (lostmsu) [inline src/TorchSharp/NN/Losses.cs:54]: Weighted loss should imho also implement `ILoss<T, TResult>`, cause the weight is optional?

File: src/TorchSharp/NN/Losses.cs [hunk 4]
```diff
@@ -12,16 +12,56 @@ namespace TorchSharp
 {
     using Modules;
 
-    public class Loss : nn.Module
+    public abstract class Loss<T1, T2, TResult> : nn.Module<T1, T2, TResult>
```
Comment (lostmsu) [inline src/TorchSharp/NN/Losses.cs:15]: Are there any losses, that do not return `Tensor`? I am curious if it makes sense to have it as a generic parameter.

File: src/TorchSharp/NN/Module.cs [hunk 5]
```diff
@@ -1120,6 +1111,59 @@ protected void Dispose(bool disposing)
                     }
                 }
             }
+
+            public interface IModule<T1, TResult>
```
Comment (lostmsu) [inline src/TorchSharp/NN/Module.cs:1115]: Nitpick which I am unsure if it makes sense: it is `Func<T,TResult>`, not `Func<T1, TResult>`.

File: src/TorchSharp/NN/Recurrent/GRU.cs [hunk 6]
```diff
@@ -25,7 +25,7 @@ internal GRU(IntPtr handle, IntPtr boxedHandle, long hiddenSize, long numLayers,
             private bool _batch_first;
 
             [DllImport("LibTorchSharp")]
-            extern static IntPtr THSNN_GRU_forward(torch.nn.Module.HType module, IntPtr input, IntPtr h_0, out IntPtr h_n);
+            extern static IntPtr THSNN_GRU_forward(torch.nn.Module<Tensor, Tensor>.HType module, IntPtr input, IntPtr h_0, out IntPtr h_n);
```
Comment (lostmsu) [inline src/TorchSharp/NN/Recurrent/GRU.cs:28]: Unrelated to this change: are there many name conflicts? Does it make sense for every TorchSharp file to have `using static torch;` and `using static torch.nn` to be concise? Maybe even as a `global using`.

File: src/TorchSharp/NN/Sequential.cs [hunk 7]
```diff
@@ -16,30 +16,32 @@ namespace Modules
         /// <summary>
         /// This class is used to represent a Sequential module.
         /// </summary>
-        public class Sequential : torch.nn.Module
+        public class Sequential : torch.nn.Module<Tensor, Tensor>
```
Comment (lostmsu) [inline src/TorchSharp/NN/Sequential.cs:19]: As I mentioned in the discussion, it would make sense for a `Sequential<T>: Module<T, T>`

File: src/TorchSharp/NN/Activation/ELU.cs (hunk 2 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Activation/ELU.cs:20]: No.

File: src/TorchSharp/NN/Losses.cs (hunk 3 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Losses.cs:54]: Not sure what you mean -- implementing `ILoss<T, TResult>` would mean that there's a `TResult forward(T);` and none of the loss functions have one like that.

File: src/TorchSharp/NN/Losses.cs (hunk 4 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Losses.cs:15]: You can imagine some that return "double" or something like that, even if there's none that currently return anything but a tensor.

File: src/TorchSharp/NN/Losses.cs (hunk 3 above)
Comment (lostmsu) [reply src/TorchSharp/NN/Losses.cs:54]: Sorry, I misread that one thinking it provides an optional parameter `weight` instead of having it on the class instance itself.

File: src/TorchSharp/NN/Recurrent/GRU.cs (hunk 6 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Recurrent/GRU.cs:28]: As you said, it's unrelated. Maybe something to fix in the future.

File: src/TorchSharp/JIT/ScriptModule.cs (hunk 1 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/JIT/ScriptModule.cs:303]: I don't think so. If we encounter that, we can revisit.

**Comment by NiklasGustafsson** [comment]
Thank you @kaiidams for demonstrating the value of using generics, and thank you @lostmsu for reviewing the draft. It's ready for review now.
